lib/rubygems/ext/cmake_builder.rb

```diff
@@ -1,21 +1,115 @@
 
+# This builder creates extensions defined using CMake. Its is invoked if a Gem's spec file
+# sets the `extension` property to a string that contains `CMakeLists.txt`.
+#
+# In general, CMake projects are built in two steps:
+#
+#  * configure
+#  * build
+#
+# The builder follow this convention. First it runs a configuration step and then it runs a build step.
+#
+# CMake projects can be quite configurable - it is likely you will want to specify options when
+# installing a gem. To pass options to CMake specify them after `--` in the gem install command. For example:
+#
+#   gem install <gem_name> -- --preset <preset_name>
+#
+# Note that options are ONLY sent to the configure step - it is not currently possible to specify
+# options for the build step. If this becomes and issue then the CMake builder can be updated to
+# support build options.
+#
+# Useful options to know are:
+#
+#  -G to specify a generator (-G Ninja is recommended)
+#  -D<CMAKE_VARIABLE> to set a CMake variable (for example -DCMAKE_BUILD_TYPE=Release)
+#  --preset <preset_name> to use a preset
+#
+# If the Gem author provides presets, via CMakePresets.json file, you will likely want to use one of them.
+# If not, you may wish to specify a generator. Ninja is recommended because it can build projects in parallel
+# and thus much faster than building them serially like Make does.
+
 class Gem::Ext::CmakeBuilder < Gem::Ext::Builder
-  def self.build(extension, dest_path, results, args=[], lib_dir=nil, cmake_dir=Dir.pwd,
-    target_rbconfig=Gem.target_rbconfig)
+  attr_accessor :runner, :profile
+
+  def initialize
+    @runner = self.class.method(:run)
+    @profile = :release
+  end
+
+  def build(extension, dest_path, results, args = [], lib_dir = nil, cmake_dir = Dir.pwd, target_rbconfig = Gem.target_rbconfig)
     if target_rbconfig.path
       warn "--target-rbconfig is not yet supported for CMake extensions. Ignoring"
     end
 
-    unless File.exist?(File.join(cmake_dir, "Makefile"))
-      require_relative "../command"
-      cmd = ["cmake", ".", "-DCMAKE_INSTALL_PREFIX=#{dest_path}", *Gem::Command.build_args]
-
-      run cmd, results, class_name, cmake_dir
+    # Make sure lib dir is set
+    unless lib_dir
+      lib_dir ||= File.join(dest_path, "lib")
     end
 
-    make dest_path, results, cmake_dir, target_rbconfig: target_rbconfig
+    # Figure the build dir
+    build_dir = File.join(cmake_dir, "build")
+
+    # Check if the gem defined presets
+    check_presets(cmake_dir, args, results)
+
+    # Configure
+    configure(cmake_dir, build_dir, lib_dir, args, results)
+
+    # Compile
+    compile(cmake_dir, build_dir, args, results)
 
     results
   end
+
+  def configure(cmake_dir, build_dir, lib_dir, args, results)
+    cmd = ["cmake",
+           cmake_dir,
+           "-B",
+           build_dir,
+           "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=#{lib_dir}", # Windows
+           "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=#{lib_dir}", # Not Windows
```

Comment:
So, if I'm understanding correctly, the `cmake --build` command that goes after this configure step is then aware of the artifacts configure is putting in this directory, correct?